crates/typst-syntax/src/parser.rs

```diff
@@ -1138,18 +1138,40 @@ fn module_import(p: &mut Parser) {
     let m = p.marker();
     p.assert(SyntaxKind::Import);
     code_expr(p);
-    if p.eat_if(SyntaxKind::Colon) && !p.eat_if(SyntaxKind::Star) {
-        import_items(p);
+    if p.eat_if(SyntaxKind::As) {
+        // allow renaming a full module import
+        p.expect(SyntaxKind::Ident);
+        p.wrap(m, SyntaxKind::ModuleImport);
+
+        // friendlier error if the user tries to rename and import items at the
+        // same time
+        if p.at(SyntaxKind::Colon) {
+            p.unexpected();
+            p.hint("cannot import the renamed module and its items at the same time");
+            p.hint("try importing the module's items in a separate `import` expression");
```

Comment:
maybe still call `import_items` here to prevent further errors for them?

Reply:
I think that could be weird, mostly because that would make the parser check if the items are valid, when this is supposed to be invalid syntax, so it would say something like "missing identifier" in some item for example when, in reality, the entire construction is supposed to be invalid (not to mention that the marker wrap, if delayed as you suggest, would also include the items, which would make even less sense for an invalid construction!).

I can add this either way if you think that's fine though.

Of course, we could also just support importing the module and its items at the same time through renaming. I honestly don't have a very strong opinion here (I made it invalid only to keep the status quo for now, as it could then be added in a future PR if we ever change our minds later). I'll be keeping such behavior invalid for now, but feel free to tell me if you have diverging opinions in this sense.

Reply:
Update: It seems that calling `import_items` here is necessary to move the marker outside the 'if' as you suggest (otherwise a more general parsing error takes priority over the colon error). I'm not sure what I prefer, but if this seems like the best path to you then I'll push it.

Reply:
> not to mention that the marker wrap, if delayed as you suggest, would also include the items, which would make even less sense for an invalid construction

I don't think this would make less sense actually. This is clearly how any human parses it and the parser should parse invalid constructions as closely to have humans interpret them, even if they don't follow the AST structure.

Reply:
> Of course, we could also just support importing the module and its items at the same time through renaming. I honestly don't have a very strong opinion here (I made it invalid only to keep the status quo for now, as it could then be added in a future PR if we ever change our minds later). I'll be keeping such behavior invalid for now, but feel free to tell me if you have diverging opinions in this sense.

For a rename it makes sort of sense to allow this. I doesn't happen by default for non-renamed imports because there'd be no way to import items without importing the module then.

Reply:
> > not to mention that the marker wrap, if delayed as you suggest, would also include the items, which would make even less sense for an invalid construction
> 
> I don't think this would make less sense actually. This is clearly how any human parses it and the parser should parse invalid constructions as closely to how humans interpret them, even if they don't follow the AST structure.

Alright. I will be pushing the changes then. :+1:

Reply:
> > Of course, we could also just support importing the module and its items at the same time through renaming. I honestly don't have a very strong opinion here (I made it invalid only to keep the status quo for now, as it could then be added in a future PR if we ever change our minds later). I'll be keeping such behavior invalid for now, but feel free to tell me if you have diverging opinions in this sense.
> 
> For a rename it makes sort of sense to allow this. I doesn't happen by default for non-renamed imports because there'd be no way to import items without importing the module then.

I guess that makes sense (Rust also lets you do something similar). Although maybe it might seem to the end user that one line is doing "too much" (like `#import "something.typ" as x: y, z as b, w`). I'm not sure if that's even a real concern, though (maybe I'm just thinking too harshly). Do you think I should implement it then (should be easy)?

Reply:
Why not

Reply:
Ok, done! :+1: